python/lsst/ip/diffim/utils.py

```diff
@@ -1080,36 +1084,115 @@ def detectDipoleSources(self, doMerge=True, diffim=None, detectSigma=5.5, grow=3
+    return np.nanmean(width) if average else width
+
+
+def evaluateMeanPsfFwhm(exposure, average=True, *, fwhmExposureBuffer=None, fwhmExposureGrid=None):
+    """Get the median PSF FWHM by evaluating it on a grid within an exposure.
+
+    Parameters
+    ----------
+    exposure : `~lsst.afw.image.Exposure`
+        The exposure for which the mean FWHM of the PSF is to be computed.
+        The exposure must contain a `psf` attribute.
+    average : `bool`, optional
+        Set to return the average width over Y and X axes.
+    fwhmExposureBuffer : `float`, optional
+        Fractional buffer margin to be left out of all sides of the image
+        during the construction of the grid to compute mean PSF FWHM in an
+        exposure.
+    fwhmExposureGrid : `int`, optional
+        Grid size to compute the mean FWHM in an exposure
 
-    def sliceWidth(image, threshold, peaks, axis):
-        vec = image.take(peaks[1 - axis], axis=axis)
-        low = np.interp(threshold, vec[:peaks[axis] + 1], np.arange(peaks[axis] + 1))
-        high = np.interp(threshold, vec[:peaks[axis] - 1:-1], np.arange(len(vec) - 1, peaks[axis] - 1, -1))
-        return high - low
-    width = (sliceWidth(image, peak/2., peakLocs, axis=0), sliceWidth(image, peak/2., peakLocs, axis=1))
-    return np.mean(width) if average else width
+    Returns
+    -------
+    medianFwhm : `float` | `tuple` [`float`]
+        The median PSF FWHM on the exposure
+
+    Notes
+    -----
+    If either of the ``fwhmExposureBuffer`` or ``fwhmExposureGrid`` parameters
+    is `None`, the PSF FWHM is computed at the average position of the PSF.
+    This behavior is same as that of `getPsfFwhm` but has a helpful error
+    message that indicates how to avoid the error from occurring.
+
+    See Also
+    --------
+    getPsfFwhm
+    """
+
+    psf = exposure.psf
+
+    if fwhmExposureBuffer is None or fwhmExposureGrid is None:
+        try:
+            fwhm = getPsfFwhm(psf, average=average)
+        except InvalidParameterError as err:
+            _LOG.error("Unable to compute PSF FWHM at average position. Set fwhmExposureBuffer and "
+                       "fwhmExposureGrid parameters to avoid this error.")
+            raise err
+        return fwhm
+
+    bbox = exposure.getBBox()
+
+    xmax, ymax = bbox.getMax()
+    xmin, ymin = bbox.getMin()
+
+    xbuffer = fwhmExposureBuffer*(xmax-xmin)
+    ybuffer = fwhmExposureBuffer*(ymax-ymin)
+
+    width = []
+    for (x, y) in itertools.product(np.linspace(xmin+xbuffer, xmax-xbuffer, fwhmExposureGrid),
+                                    np.linspace(ymin+ybuffer, ymax-ybuffer, fwhmExposureGrid)
+                                    ):
+        pos = geom.Point2D(x, y)
+        try:
+            fwhm = getPsfFwhm(psf, average=average, position=pos)
+            width.append(fwhm)
+        except Exception:
+            pass
+
+    if not width:
+        return np.nan if average else (np.nan, np.nan)
```

Comment:
I'm open to suggestions on whether we should return `NaN`s  here or raise an exception, as it would have been the case before. I don't know if the downstream processes are equipped to handle `NaN`s.

Reply:
If this raises an exception, it needs to be caught by the caller.  This should be much more robust to failures than before, but I really don't want to be building in more crashing opportunities.

Reply:
Remove the support for the `average` keyword. This should raise an exception rather than returning `NaN`, since downstream code can't handle `NaN`s.

Reply:
I think a `ValueError` would be appropriate if the final result would be `NaN`